### PR TITLE
Dynamic ports

### LINKED ISSUE #60: SLAPI API Plugin: Dynamic Port Assignment
- AS A: User/Developer

  - I NEED: SLAPI to assign ports to managed API plugins for internal communication (leave external ports for webhooks or alert services configurable)
  - SO THAT: Users don't have to be concerned about port management or conflicts.

### PR DESCRIPTION
**Pull request by TheFynx**
Fixes #60

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,16 +1,39 @@
...
+- export $SLAPI_VERSION=$(bundle exec gem list slapi | grep -o '[0-9]\.[0-9]\.[0-9]')
+- export GIT_TAG=$SLAPI_VERSION.b$TRAVIS_BUILD_NUMBER
+- git tag $GIT_TAG -a -m "SLAPI Build Bot: Generated tag from TravisCI for build $TRAVIS_BUILD_NUMBER"
+- git push -q https://slapibot:$GH_TOKEN@github.com/imperiallabs/slapi --tags
+- ls -R
+deploy:
+  provider: releases
+  prerelease: true
+  file:
+    - Dockerfile
```
Comment (aaronblythe) [inline .travis.yml:25]: To remain consistent, should the things in this list also be moved left?

File: lib/plugins/helpers/docker.rb [hunk 2]
```diff
@@ -100,6 +99,25 @@ def hash_set(filename = nil, script = nil)
     end
   end
 
+  def expose
+    container_port = @container_hash[:ExposedPorts].keys[0].to_s
+    ip = @config['plugin']['public'] ? '0.0.0.0' : Socket.ip_address_list.detect(&:ipv4_private?).ip_address
+    @port = @config['plugin']['exposed_port'].blank? ? @dynamic_port : @config['plugin']['exposed_port']
```
Comment (aaronblythe) [inline lib/plugins/helpers/docker.rb]: There are a lot of instance variables that likely should be attr_reader, attr_writer, or attr_accessor as appropriate.  Setting instance variables for the first time outside of the initialize is dangerous as complexity is added to code.

I believe this `port` is one of these cases.

File: lib/plugins/plugin.rb [hunk 3]
```diff
@@ -19,7 +20,7 @@ class Plugin
   attr_reader :help, :type
   attr_accessor :config
 
-  def initialize(file, settings)
+  def initialize(file, settings, port)
```
Comment (aaronblythe) [inline lib/plugins/plugin.rb]: There is a lot of indirection going on here to set port->@dynamic_port->then later to set @port. Is there possibly a more straight forward way to do this?

File: lib/plugins/plugins.rb [hunk 4]
```diff
@@ -32,8 +32,10 @@ def load
     file_location = @settings.plugins['location'] ? @settings.plugins['location'] : '../../config/plugins/'
 
     yaml_files = File.expand_path(file_location + '*.yml', File.dirname(__FILE__))
+    port = 48130
```
Comment (aaronblythe) [inline lib/plugins/plugins.rb]: Just for my understanding is this supposed to be hard coded?

Comment (TheFynx) [reply lib/plugins/plugins.rb]: We have to provide a starting point to iterate from, you'll notice there is a `port += 1` below this.

That way every plugin loaded is assigned a unique port (e.g.; 48130, 48131, 48132)

File: .travis.yml (hunk 1 above)
Comment (TheFynx) [reply .travis.yml:25]: I just moved it all to the right, I didn't realize the it took the indents away when I re-setup travis.

File: .vscode/launch.json [hunk 5]
```diff
@@ -13,7 +13,7 @@
                 "-p", "4567"
             ],
             "useBundler": true,
-            "pathToBundler": "${env.HOME}/.rvm/gems/${env.rvm_ruby_string}/wrappers/bundle",
+            "pathToBundler": "${env:HOME}/.rvm/gems/${env:rvm_ruby_string}/wrappers/bundle",
```
Comment (aaronblythe) [inline .vscode/launch.json:16]: Awesome find, this has been bugging me.

File: lib/plugins/plugins.rb [hunk 6]
```diff
@@ -32,8 +32,10 @@ def load
     file_location = @settings.plugins['location'] ? @settings.plugins['location'] : '../../config/plugins/'
 
     yaml_files = File.expand_path(file_location + '*.yml', File.dirname(__FILE__))
+    dynamic_port = 48130
```
Comment (aaronblythe) [inline lib/plugins/plugins.rb:35]: Can log a separate issue, but I feel like there should be a check here if these ports are already taken by something else (potentially even a previously loaded container).

**Review by aaronblythe — APPROVED** [review]

**Pull request merged**